### PR TITLE
[Preview] Guzzle 6, PSR 7, and Middleware Support

### PR DESCRIPTION
**Pull request by jeremeamia**
The entire guzzlehttp/command package has been rewritten to support Guzzle 6, PSR 7, Promises, and Middleware. Check it out.

### CODE REVIEW

**Comment by willwh** [comment]
This definitely gets a :+1: from me!

**Comment by mtdowling** [comment]
This is awesome! I'll try to find some time soon to take a deeper look and do a code review.

**Comment by neeckeloo** [comment]
:+1: Good job. I hoped these changes.

**Comment by mcrumm** [comment]
:+1: Looking forward to seeing this merged!

File: src/ServiceClient.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    public function executeAsync(CommandInterface $command)
+    {
+        $stack = $command->getHandlerStack() ?: $this->handlerStack;
+        $handler = $stack->resolve();
+
+        return $handler($command);
+    }
+
+    public function batch($commands, array $options = [])
```
Comment (mtdowling) [inline src/ServiceClient.php]: I think `batch` is a misnomer. Maybe `executeEach` would be more appropriate.

File: src/ServiceClient.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        // Execute the batch synchronously, then sort and return the results.
+        return $this->batchAsync($commands, $options)
+            ->then(function () use (&$results) {
+                ksort($results);
+                return $results;
+            })
+            ->wait();
+    }
+
+    public function batchAsync($commands, array $options = [])
```
Comment (mtdowling) [inline src/ServiceClient.php]: `executeEachAsync`?

**Comment by mtdowling** [comment]
This looks great. My biggest open question is how composable are `transformers`? The middleware system is being used here to transfer commands and results, but I'm wondering how users can do things like add a header to the request that will ultimately be created? It seems like that would require decoration of the transformer functions, which could then be implemented with an additional middleware stack.

In the AWS SDK for PHP (you've probably heard of it :)) we had a way to modify the request as it's being built-up: https://github.com/aws/aws-sdk-php/blob/master/src/HandlerList.php. How would you envision similar functionality being implemented here?

**Comment by jeremeamia** [comment]
The intention here is that that you still configure request-level middleware on the underlying HTTP client. This can be done prior to instantiating the command client or by calling `$serviceClient->getHttpClient()->getHandlerStack()` afterwards. The transformers should not need to be composed. I like this setup because there is separation of concerns between each layer.

- Command-level middleware are configured on service client and are concerned only with commands and results.
- Transformers are configured on the service client and are what turn commands into requests on the way in and responses into results on the way out.
- Request-level middleware are configured on the HTTP client and are concerned only with requests and responses.

Does this make sense to you now? Do you think this approach is fine, or should I consider something else?

**Comment by jeremeamia** [comment]
@mtdowling Essentially this:

```php
$serviceClient = new ServiceClient($httpClient, $commandToRequestFn, $responseToResultFn);
$serviceClient->getHandlerStack()->push($commandMiddleware);
$serviceClient->getHttpClient()->getHandlerStack()->push($httpMiddleware);
```

**Comment by csarrazi** [comment]
Small question. Why use the `HandlerStack` here? Obviously, middleware will be different between the Http client (with his own handlerStack), and the Service client, so they won't be interchangeable.

We should differentiate a command middleware from a http middleware, especially since the objects which will be passed will be different (and they won't be interchangeable).

**Comment by jeremeamia** [comment]
@mtdowling Not sure if you'd had a chance to take another look and if I answered your question. I'm not sure why you think "batch" is a misnomer either. Want to talk about that sometime?

@csarrazi I realize the command middleware and http middleware are different, and I've accounted for that. The HandlerStack object, if you ignore the `__invoke` method, is not specific to either, it just deals with callables. Basically, the HTTP client has a HandlerStack and the Command client has a HandlerStack.

File: src/ServiceClient.php (hunk 1 above)
Comment (jeremeamia) [inline src/ServiceClient.php]: We used to call is `executeAll`. Should we stick with that?

**Comment by mtdowling** [comment]
All of your reasoning and explanations sound good to me.

Whenever you think this is ready, :shipit:

**Comment by csarrazi** [comment]
@jeremeamia No problem for me. I was just afraid of misuse by a few people. Regardless, this looks good to me! :smiley:

**Comment by dccampbell** [comment]
Is this ready-to-merge or is there still some work left to be done on it? (cc @jeremeamia)

**Comment by jeremeamia** [comment]
Thanks for the nudge. I'll see if I have time to look at it this weekend.
It should be ready to go, but I'm sure there are a few docblocks or
something that needs polishing.

--
*Jeremy Lindblom (@jeremeamia <https://twitter.com/jeremeamia>)*
Software/Platform Engineer at Engrade <https://engrade.com/> (part of
McGraw-Hill
Education <http://www.mheducation.com/>)
Co-founder of the Pacific Northwest PHP Conference <http://pnwphp.com/> (
@PNWPHP <https://twitter.com/pnwphp>)
Co-author of the AWS SDK for PHP <http://aws.amazon.com/sdkforphp/> (
@awsforphp <https://twitter.com/awsforphp>)
PHP-FIG <http://www.php-fig.org/> Representative for the Guzzle
<http://guzzlephp.org/> project

Looking for a senior-level software engineering position doing PHP? I have
some available with McGraw-Hill Education in Seattle and Los Angeles (Santa
Monica or Irvine). Contact me for me more information.


On Tue, Jan 26, 2016 at 1:43 PM, David C. Campbell <notifications@github.com

File: src/HasDataTrait.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * The methods in the class are primarily for implementing the ArrayAccess,
+ * Countable, and IteratorAggregate interfaces.
+ */
+trait HasDataTrait
+{
+    /** @var array Data stored in the collection. */
+    protected $data;
+
+    public function __toString()
```
Comment (snoek09) [inline src/HasDataTrait.php]: add missing PHPDoc for methods

File: src/ResultInterface.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?php
+namespace GuzzleHttp\Command;
+
+/**
+ * @TODO description
```
Comment (snoek09) [inline src/ResultInterface.php]: Add description

**Comment by saintberry** [comment]
+1 for this fix. We're keen to upgrade to php7 and are waiting on this getting merged.

File: src/HasDataTrait.php (hunk 3 above)
Comment (jeremeamia) [inline src/HasDataTrait.php]: `__toString` doesn't really need a docblock since you don't call it directly.

**Comment by jeremeamia** [comment]
This has been merged now and the previous Guzzle 5 stuff will continue to exist in the 0.8 branch. I still need to update the CHANGELOG and README before I can tag version 0.9.0.

**Pull request merged**